aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -198,34 +166,16 @@ static ScalarType get_dtype(Tensor& result, const Tensor& self, optional<ScalarT
   return src_type;
 }
 
-static Tensor& sum_out(Tensor& result, const Tensor& self, IntList dim,
-                       bool keepdim, optional<ScalarType> opt_dtype) {
-  ScalarType dtype = get_dtype(result, self, opt_dtype, true);
-  auto iter = make_reduction("sum", result, self, dim, keepdim, dtype);
-  if (iter->numel() == 0) {
-    result.zero_();
-  } else {
-    sum_stub(iter->device_type(), *iter);
-  }
-  return result;
-}
-
-static Tensor sum(const Tensor& self, IntList dim, bool keepdim, optional<ScalarType> dtype) {
-  Tensor result;
-  native::sum_out(result, self, dim, keepdim, dtype);
-  return result;
-}
-
-Tensor sum(const Tensor &self, ScalarType dtype) {
-  return at::native::sum(self, {}, false, optional<ScalarType>(dtype));
+Tensor mean(const Tensor& self, optional<ScalarType> dtype) {
+  return at::native::mean(self, {}, false, dtype);
 }
 
-Tensor sum(const Tensor &self) {
-  return at::native::sum(self, {}, false, c10::nullopt);
+Tensor sum(const Tensor &self, optional<ScalarType> dtype) {
+  return at::native::sum(self, {}, false, dtype);
 }
 
-static Tensor& prod_out(Tensor& result, const Tensor& self, IntList dim,
-                        bool keepdim, optional<ScalarType> opt_dtype) {
+static Tensor& prod_impl(Tensor& result, const Tensor& self, IntList dim,
```

Comment:
I don't really understand the organization here.  It's not grouped by function name, which makes it difficult to figure out how a specific function is implemented.  It's also not really grouped by implementation type, i.e. some functions have fn_impl, some fn_out, etc.  I'd just go back to function name grouping.

Reply:
Why is there a prod_impl but not a sum_impl?  What's the distinction?

Reply:
Because the signatures are different:

- `prod(Tensor self, int64_t dim, ...`
- `sum(Tensor self, IntList[1] dim, ...`

One of the sum native impl can be re-used by othes but prod wasn't so lucky.

Let me take a stab at changing the prod signature to use `IntList[1]`. This should be backwards compatible.

Reply:
Actually, the current organization makes sense in its own way. `cumsum` and `cumprod` are not really reduce functions, they are cumulative.

I'll still change to order by function, I already spent way too much mental effort for grasping what's going on here.

Reply:
> Let me take a stab at changing the prod signature to use IntList[1]. This should be backwards compatible.

Punting on this because `prod_backward` needs to be rewritten to support multi-dim. This needs to be a separate PR.

Reply:
I would recommend ordering by what you think makes sense (with an explanation); I was just describing why the current ordering didn't make sense to me.